### ISSUE TITLE
Structured logging support (JSON) in console and log file

### ISSUE THREAD

**Issue by SadiHassan** [ClickHouse/ClickHouse#35366]
**Use case**

This feature will provide JSON formatted log output. The purpose is to allow easier ingestion and query in log analysis tools.

**Describe the solution you'd like**
1. A sample output might look like this.
```{"DATE_TIME": "2022.03.10 15:46:19.000418", "THREAD": "408576", "QUERY": "", "PRIORITY": "Trace", "SRC": "AsynchronousMetrics", "MSG": "MemoryTracking: was 311.37 MiB, peak 319.11 MiB, will set to 315.43 MiB (RSS), difference: 4.06 MiB"}```
2. Both console and file will have such output.
3. User will be able to change names of the keys using `config.xml` file.
4. User will be able to turn on/off `JSON` output and will have ability to choose which keys to output and which not to output. 
For example consider the sample tag below (that will be written in `config.xml`). 

```
<json>
    <datetime>DATE_TIME</datetime>
    <thread_id> THREAD </thread_id>
    <query_id> QUERY </query_id>
    <priority> PRIORITY </priority>
    <source>SRC</source>
    <message>MSG</message>
</json>
```
If user deletes or comments out the whole `json` tag, usual log output will prevail. User can change values of any keys from the tag and also can delete or comments out any key if they don't want to see that key(s) in output.

**Comment by den-crane** [comment]
duplicate of https://github.com/ClickHouse/ClickHouse/issues/20130  https://github.com/ClickHouse/ClickHouse/issues/12300

**Comment by den-crane** [comment]
>Please let me know if you are good for me to work with the proposal I gave and later on yours

Basically, you don't need anyone's permission.

**Comment by SadiHassan** [comment]
> > Please let me know if you are good for me to work with the proposal I gave and later on yours
> 
> Basically, you don't need anyone's permission, especially mine, because I am doing triage.

Thanks @den-crane ! I wanted to make sure if we are on the same page and if you are good with the idea. Because, later on after submitting the PR if we find the direction was wrong, that will be a waste of time!  
So, I will work on this and submit a PR when I am done. Basically, I will add a `<format>` section in config (under `logger`) that will allow user to format log by their choice (for now, `json` only). Also, if it is `json`, user will be able to choose keys and select/deselect (basically delete or comment out) any key(s). User will also be able to write their own key-names. (for example, if user wants to see `thread_id`s as `TID`, they will modify config under `logger.json` as `<thread_id>TID</thread_id>`. I will implement these both on console and files. Once json is done, I plan to work on generating CSV and TSV log files. Sounds good?

**Comment by den-crane** [comment]
I don't really understand why do you need to map tags.

Are you aware that you can enable system.text_log and query it?

**Comment by SadiHassan** [comment]
Yes, I am aware of that. But I want to output json format logs in console. Also, the mapping is required so that we can change the key names while needed to be used by log analytics in real time.

**Comment by alexey-milovidov** [comment]
Let's make a simple switch to JSON format without any more settings. The field names will be predefined. All possible fields will be present. This will include file/line that are missing in the usual logs. The strings will be validated for UTF-8 with invalid data replaced to �.

**Comment by SadiHassan** [comment]
> This will include file/line that are missing in the usual logs. 
Can you give any example? I am not sure how to find what's missing in usual logs!
>The strings will be validated for UTF-8 with invalid data replaced to �.
String can be validated but this might effect the log printing efficiency to the console. Need to check.

FYI: A PR is submitted with simple switch option to JSON or no JSON: https://github.com/ClickHouse/ClickHouse/pull/35694

**Comment by alexey-milovidov** [comment]
> String can be validated but this might effect the log printing efficiency to the console.

No way it can visibly affect performance if implemented correctly.

**Comment by alexey-milovidov** [comment]
See also WriteHelpers.h, writeJSONString.

**Comment by UnamedRus** [comment]
It can be closed, i think
